[InstCombine] Fold shift+cttz with power of 2 operands

MDevereau opened the pull request:
#121386 Introduced cttz intrinsics which caused a regression where vscale/vscale divisions could no longer be constant folded.

This fold was suggested as a fix in #126411. https://alive2.llvm.org/ce/z/gWbtPw

MDevereau committed:
[InstCombine] Fold cttz with power of 2 operands

(#121386) Introduced cttz intrinsics which caused a regression
where vscale/vscale divisions could no longer be constant folded.

This fold was suggested as a fix in (#126411)

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1613,6 +1613,22 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+  if (match(Op0, m_Shl(m_Value(Shl0_Op0), m_Value(Shl0_Op1))) &&
+      match(Op1, m_Intrinsic<Intrinsic::cttz>(m_BinOp(Shl1))) &&
+      match(Shl1, m_Shl(m_Value(Shl1_Op0), m_Value(Shl1_Op1))) &&
+      isKnownToBeAPowerOfTwo(Shl1, false, 0, SQ.getWithInstruction(&I).CxtI) &&
```
```suggestion
      isKnownToBeAPowerOfTwo(Shl0_Op0, /*OrZero=*/true, 0, &I) &&
```
It also holds for zero: https://alive2.llvm.org/ce/z/UnGp5a

MDevereau replied:
Done

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1613,6 +1613,22 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+  BinaryOperator *Shl1;
+  if (match(Op0, m_Shl(m_Value(Shl0_Op0), m_Value(Shl0_Op1))) &&
+      match(Op1, m_Intrinsic<Intrinsic::cttz>(m_BinOp(Shl1))) &&
+      match(Shl1, m_Shl(m_Value(Shl1_Op0), m_Value(Shl1_Op1))) &&
+      isKnownToBeAPowerOfTwo(Shl1, false, 0, SQ.getWithInstruction(&I).CxtI) &&
+      Shl0_Op0 == Shl1_Op0) {
```
```suggestion
      match(Shl1, m_Shl(m_Specific(Shl0_Op0), m_Value(Shl1_Op1))) &&
      isKnownToBeAPowerOfTwo(Shl1, false, 0, SQ.getWithInstruction(&I).CxtI)) {
```

MDevereau replied:
Done

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1613,6 +1613,22 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+      Shl0_Op0 == Shl1_Op0) {
+    auto *Shl0 = cast<BinaryOperator>(Op0);
+    if ((Shl0->hasNoUnsignedWrap() && Shl1->hasNoUnsignedWrap()) ||
+        (Shl0->hasNoSignedWrap() && Shl1->hasNoSignedWrap())) {
```
Proof for nsw?

MDevereau replied:
Sorry if I've misunderstood your question,
If you're asking for alive test I think your comment from earlier proves it: https://alive2.llvm.org/ce/z/gWbtPw
If you're asking for an IR test, InstCombine sets nsw/nuw flags very early on for shift binops, so a separate IR test for both nuw and nsw are actually testing both nsw and nuw

dtcxzyw replied:
> If you're asking for an alive test I think your comment from earlier proves it: https://alive2.llvm.org/ce/z/gWbtPw

Please add it to the PR description.

MDevereau replied:
Done

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1613,6 +1613,22 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+    if ((Shl0->hasNoUnsignedWrap() && Shl1->hasNoUnsignedWrap()) ||
+        (Shl0->hasNoSignedWrap() && Shl1->hasNoSignedWrap())) {
+      Value *NewShl =
+          Builder.CreateShl(ConstantInt::get(Shl1->getType(), 1), Shl0_Op1);
```
We can propagate nowrap flags to shl: https://alive2.llvm.org/ce/z/XBcBmZ

MDevereau replied:
Done

dtcxzyw commented on llvm/test/Transforms/InstCombine/shift-cttz-ctlz.ll:
```diff
@@ -103,4 +103,38 @@ entry:
+; CHECK-NEXT:    ret i64 4
+;
+entry:
+  %0 = tail call i64 @llvm.vscale.i64()
```
Use named values.

MDevereau replied:
Done

dtcxzyw commented on llvm/test/Transforms/InstCombine/shift-cttz-ctlz.ll:
```diff
@@ -103,4 +103,38 @@ entry:
+  ret i32 %div1
+}
+
+declare i64 @llvm.vscale.i64()
+declare i64 @llvm.cttz.i64(i64, i1 immarg)
+declare i32 @llvm.vscale.i32()
+declare i32 @llvm.cttz.i32(i32, i1 immarg)
```
Intrinsic declarations are not needed.

MDevereau replied:
Removed

MDevereau committed:
Add fold for zero
Propagate nowrap flags to new shift
Use named values in tests
Remove intrinsic declarations

MDevereau committed:
Remove unnecessary line

dtcxzyw reviewed (approved):
LGTM.
